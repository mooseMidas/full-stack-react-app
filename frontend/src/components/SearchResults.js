/*
SearchResults component will render results list,
"ArrowDown" button to add item to Favourites component, 
"Preview" button to autostart previews of item result
*/

import React, { useState } from "react";
import { Container, Row, Col } from "react-bootstrap";
import { ArrowDown, XCircle } from "react-bootstrap-icons";

// receives three props from App.js -> to display and update the list and to set the favourites list
function SearchResults({ searchResults, favourites, setFavourites }) {
  const [previewUrl, setPreviewUrl] = useState("");

  // checks if the item has been added already using some(). If isDuplicate value is false, item is added to favourites
  const handleAddToFavourites = (result) => {
    const isDuplicate = favourites.some((favourite) => {
      const uniqueKey = result.trackId || result.collectionId;
      return (
        uniqueKey && uniqueKey === (favourite.trackId || favourite.collectionId)
      );
    });

    if (isDuplicate) {
      window.alert("This item is already in your favourites!");
    } else {
      setFavourites([...favourites, result]);
      window.alert("This item has been added to your favourites!");
    }
  };

  const handlePreview = (url) => {
    setPreviewUrl(url);
  };

  const handleClosePreview = () => {
    setPreviewUrl("");
  };

  return (
    <Container fluid="md">
      <Row>
        {searchResults.map((result) => (
          <Col
            md
            className="card-col"
            key={result.trackId || result.collectionId}
          >
            <div className="card-item">
              <div className="card--image-container">
                <img
                  src={result.artworkUrl100}
                  className="card--image"
                  alt={result.collectionName || result.trackName}
                ></img>
                <div className="card--button-container">
                  <ArrowDown
                    className="card--button"
                    onClick={() => handleAddToFavourites(result)}
                    data-testid="add-to-favourites-button"
                  />
                </div>
              </div>
              <h3 className="card--title">
                {result.collectionName || result.trackName}
              </h3>
              <div className="card-description-container">
                <span className="card--description">
                  {result.shortDescription ||
                    result.description ||
                    result.longDescription}
                </span>
              </div>
              <h3 className="card--title">
                {result.kind || result.wrapperType}
              </h3>
              {/* Preview btn will only be displayed if result has previewUrl key */}
              {result.previewUrl && (
                <button
                  className="card--preview-btn"
                  onClick={() => handlePreview(result.previewUrl)}
                >
                  Preview
                </button>
              )}
            </div>
          </Col>
        ))}
        {previewUrl && (
          <div className="preview-content">
            <video controls autoPlay className="preview">
              <source src={previewUrl} type="video/mp4" />
            </video>
            <XCircle className="close-preview-btn" onClick={handleClosePreview}/>
          </div>
        )}
      </Row>
    </Container>
  );
}

export default SearchResults;
