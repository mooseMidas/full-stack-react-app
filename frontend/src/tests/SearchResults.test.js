import React from "react";
import renderer from "react-test-renderer";
import SearchResults from "../components/SearchResults"

// Mock the setFavourites function
const mockSetFavourites = jest.fn();

// Sample search results for testing
const searchResults = [
  {
    trackId: 1,
    collectionName: "Album 1",
    shortDescription: "Short description 1",
    previewUrl: "https://example.com/video1.mp4",
  },
  {
    trackId: 2,
    collectionName: "Album 2",
    shortDescription: "Short description 2",
    previewUrl: "https://example.com/video2.mp4",
  },
];

// Sample favourites list for testing
const favourites = [
  {
    trackId: 3,
    collectionName: "Album 3",
    shortDescription: "Short description 3",
  },
];

test("SearchResults component renders correctly", () => {
  const component = renderer.create(
    <SearchResults
      searchResults={searchResults}
      favourites={favourites}
      setFavourites={mockSetFavourites}
    />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
