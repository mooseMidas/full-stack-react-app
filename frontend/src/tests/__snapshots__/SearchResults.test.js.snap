// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults component renders correctly 1`] = `
<div
  className="container-md"
>
  <div
    className="row"
  >
    <div
      className="card-col col-md"
    >
      <div
        className="card-item"
      >
        <div
          className="card--image-container"
        >
          <img
            alt="Album 1"
            className="card--image"
          />
          <div
            className="card--button-container"
          >
            <svg
              className="card--button"
              data-testid="add-to-favourites-button"
              fill="currentColor"
              height="1em"
              onClick={[Function]}
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 1a.5.5 0 0 1 .5.5v11.793l3.146-3.147a.5.5 0 0 1 .708.708l-4 4a.5.5 0 0 1-.708 0l-4-4a.5.5 0 0 1 .708-.708L7.5 13.293V1.5A.5.5 0 0 1 8 1z"
                fillRule="evenodd"
              />
            </svg>
          </div>
        </div>
        <h3
          className="card--title"
        >
          Album 1
        </h3>
        <div
          className="card-description-container"
        >
          <span
            className="card--description"
          >
            Short description 1
          </span>
        </div>
        <h3
          className="card--title"
        />
        <button
          className="card--preview-btn"
          onClick={[Function]}
        >
          Preview
        </button>
      </div>
    </div>
    <div
      className="card-col col-md"
    >
      <div
        className="card-item"
      >
        <div
          className="card--image-container"
        >
          <img
            alt="Album 2"
            className="card--image"
          />
          <div
            className="card--button-container"
          >
            <svg
              className="card--button"
              data-testid="add-to-favourites-button"
              fill="currentColor"
              height="1em"
              onClick={[Function]}
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 1a.5.5 0 0 1 .5.5v11.793l3.146-3.147a.5.5 0 0 1 .708.708l-4 4a.5.5 0 0 1-.708 0l-4-4a.5.5 0 0 1 .708-.708L7.5 13.293V1.5A.5.5 0 0 1 8 1z"
                fillRule="evenodd"
              />
            </svg>
          </div>
        </div>
        <h3
          className="card--title"
        >
          Album 2
        </h3>
        <div
          className="card-description-container"
        >
          <span
            className="card--description"
          >
            Short description 2
          </span>
        </div>
        <h3
          className="card--title"
        />
        <button
          className="card--preview-btn"
          onClick={[Function]}
        >
          Preview
        </button>
      </div>
    </div>
  </div>
</div>
`;
